### PR TITLE
Form & TextInput

### PR DESCRIPTION
**Pull request by jonnystening**
This branch comprises of the `form` and `text-input` packages

- The `form` package provides a number of components which came across from GUI2.0's `GUI-form-base`

- The `text-input` package provides the `TextInput` component (which was GUI2.0's `GUI-input-fields`).

Ref https://github.com/WestpacGEL/GEL/issues/57

### CODE REVIEW

File: components/form/src/FormGroup.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+// ==============================
+
+const asArray = val => (Array.isArray(val) ? val : [val]);
+
+// ==============================
+// Component
+// ==============================
+
+export const FormGroup = ({ primary, ...props }) => {
+	const { breakpoints, form } = useTheme();
```
Comment (jaortiz) [inline components/form/src/FormGroup.js]: can destructure further to clean up long object variables
```
const { breakpoints, form: { group } = useTheme();
```

File: components/form/src/FormLabel.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+// ==============================
+// Utils
+// ==============================
+
+// ==============================
+// Component
+// ==============================
+
+export const FormLabel = ({ sublabel, tag: Tag, htmlFor, ...props }) => {
+	const { form } = useTheme();
```
Comment (jaortiz) [inline components/form/src/FormLabel.js]: destructure further
```
const { form: { label }} = useTheme()
```

File: components/form/src/FormSection.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+// ==============================
+// Utils
+// ==============================
+
+// ==============================
+// Component
+// ==============================
+
+export const FormSection = ({ noPadding, ...props }) => {
+	const { breakpoints, form } = useTheme();
```
Comment (jaortiz) [inline components/form/src/FormSection.js]: destructure futher
```
const { breakpoints, form: { section }} = useTheme();
```

File: components/form/src/InputClusterItem.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+// ==============================
+// Utils
+// ==============================
+
+// ==============================
+// Component
+// ==============================
+
+export const InputClusterItem = ({ horizontal, ...props }) => {
+	const { form } = useTheme();
```
Comment (jaortiz) [inline components/form/src/InputClusterItem.js]: destructure further

File: components/form/src/Fieldset.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+export const Fieldset = ({ legend, children, ...props }) => {
+	const { spacing } = useContext(FormContext);
+
+	const common = {
+		border: 'none',
+		margin: 0,
+		padding: 0,
+
+		// TODO candidate for a global reset in Core? (as with GUI2.0)
```
[the remark this answers is not in the document]
Comment (jonnystening) [reply components/form/src/Fieldset.js]: Already in the reset in Core. Got rid of this

**Comment by jonnystening** [comment]
> Remove util comment if unused and remove 'Defaults to' comment across all the components

Did you look at the text-input component too? It's also part of this PR

File: components/form/src/FormSection.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	const mq = paint(breakpoints);
+
+	const common = {
+		position: 'relative', //for `.form-section-actions` positioning
+		paddingTop: section.paddingTop,
+		paddingBottom: section.paddingBottom, //6px assuming there will be a `.form-group` margin-bottom (30px)
+		paddingLeft: noPadding ? null : section.paddingLeft,
+		paddingRight: noPadding ? null : section.paddingRight,
+
+		// TODO apply programatically? (rather than via CSS pseudo-class)
```
Comment (jaortiz) [inline components/form/src/FormSection.js]: Did you decide on how to do this? I think pseudo class is fine

File: components/form/src/FormLabel.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+// Component
+// ==============================
+
+export const FormLabel = ({ sublabel, tag: Tag, htmlFor, srOnly, ...props }) => {
+	const {
+		form: { label },
+	} = useTheme();
+	const { spacing } = useContext(FormContext);
+
+	/* if (srOnly) {
```
Comment (jaortiz) [inline components/form/src/FormLabel.js]: If unused remove

File: components/form/src/FormGroup.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	const {
+		breakpoints,
+		form: { group },
+	} = useTheme();
+	const { spacing, inline } = useContext(FormContext);
+	const mq = paint(breakpoints);
+
+	const common = {
+		display: inline ? [null, 'inline-block'] : null,
+		verticalAlign: inline ? [null, 'middle'] : null,
```
Comment (jaortiz) [inline components/form/src/FormGroup.js]: refactor to `inline && ...`

File: components/form/src/FormGroup.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+		marginBottom: inline
+			? [(mb => (Array.isArray(mb) ? mb[0] : mb))(group.spacing[spacing].marginBottom), 0]
+			: group.spacing[spacing].marginBottom,
+		textAlign: primary ? 'center' : null,
```
Comment (jaortiz) [inline components/form/src/FormGroup.js]: `textAlign: primary && 'center'

File: components/form/src/FormSection.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+	const {
+		breakpoints,
+		form: { section },
+	} = useTheme();
+	const mq = paint(breakpoints);
+
+	const common = {
+		position: 'relative', //for `.form-section-actions` positioning
+		paddingLeft: noPadding ? null : section.paddingLeft,
+		paddingRight: noPadding ? null : section.paddingRight,
```
Comment (jaortiz) [inline components/form/src/FormSection.js]: `!noPadding && section.padding`

File: components/form/src/InputCluster.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+	// Pass the selected prop on to children (that way InputClusterItem styling can be set by parent InputCluster)
+	const giftedChildren = Children.map(children, child => {
+		return React.isValidElement(child) ? cloneElement(child, { horizontal }) : child;
+	});
+
+	return (
+		<div
+			css={{
+				display: horizontal ? 'flex' : null,
+				flexWrap: horizontal ? 'wrap' : null,
```
Comment (jaortiz) [inline components/form/src/InputCluster.js]: `flexWrap: horizontal && 'wrap'`

File: components/form/src/InputClusterItem.js [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+export const InputClusterItem = ({ horizontal, ...props }) => {
+	const {
+		form: { cluster },
+	} = useTheme();
+
+	const common = {
+		marginRight: horizontal ? cluster.item.horizontal.marginRight : null,
+		display: horizontal ? 'flex' : null,
+		flexDirection: horizontal ? 'column' : null,
+		justifyContent: horizontal ? 'flex-end' : null,
```
Comment (jaortiz) [inline components/form/src/InputClusterItem.js]: refactor to &&

File: components/form/src/InputClusterItem.js [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+		// Subequent items
+		'& + &': {
+			marginTop: !horizontal ? cluster.item.default.marginTop : null,
+		},
```
Comment (jaortiz) [inline components/form/src/InputClusterItem.js]: refactor to &&

File: components/text-input/src/TextInput.js [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+			'@media print': {
+				backgroundColor: 'transparent',
+
+				':disabled': {
+					backgroundColor: '#fff',
+				},
+			},
+		},
+		textarea: {
+			verticalAlign: inline ? 'top' : null,
```
Comment (jaortiz) [inline components/text-input/src/TextInput.js]: refactor to &&

File: components/text-input/src/TextInput.js [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+		<Tag
+			css={mq({
+				...styleCommon,
+				...styleAppearance[Tag],
+				...styleFixedWidth(),
+			})}
+			type={Tag === 'input' ? 'text' : undefined}
+			{...props}
+		>
+			{Tag === 'select' ? children : null}
```
Comment (jaortiz) [inline components/text-input/src/TextInput.js]: refactor to &&

Comment (jonnystening) [reply components/text-input/src/TextInput.js]: I need to return `null` explicitly in this case or it errors (`<input type="text" />` and `<input type="textarea" />`) can't receive children in React).

**Pull request merged**